client/src/hooks/useSpotify.js

```diff
@@ -0,0 +1,52 @@
+import { useEffect } from 'react'
+import { useDataLayerValue } from '../DataLayer'
+import SpotifyWebApi from 'spotify-web-api-js'
+import getTokenFromUrl from '../services/getTokenFromUrl'
+
+export default function useToken() {
+  const spotify = new SpotifyWebApi()
+  const [{ token, track }, dispatch] = useDataLayerValue()
+
+  useEffect(() => {
+    const hash = getTokenFromUrl()
+    window.location.hash = ''
+    const _token = hash.access_token
+
+    if (_token) {
+      dispatch({
+        type: 'SET_TOKEN',
+        token: _token,
+      })
+
+      spotify.setAccessToken(_token)
+
+      spotify
+        .searchTracks(getRandomCharacter(), {
+          limit: 15,
+          offset: getRandomOffset(),
+          market: 'DE',
+        })
+        .then((track) =>
+          dispatch({
+            type: 'SET_TRACK',
+            track: track,
+          })
+        )
+        .catch((error) => {
+          console.error(error)
+        })
+    }
+
+    function getRandomCharacter() {
+      const characters = 'abcdefghijklmnopqrstuvwxyz'
+      return characters.charAt(Math.floor(Math.random() * characters.length))
+    }
+
+    function getRandomOffset() {
+      return Math.floor(Math.random() * 500)
+    }
```

Comment:
```suggestion
    const characters = 'abcdefghijklmnopqrstuvwxyz'

    function getRandomCharacter(rng=Math.random) {  
      return characters.charAt(Math.floor(rng() * characters.length))
    }

    function getRandomOffset(limit=500, rng=Math.random) {
      return Math.floor(rng() * limit)
    }
```

These functions can be safely moved to their own file, they have no dependency and don't need to be redefined on every update. 

Same goes for the the `characters` string, it will be constant over the lifetime of your app.

Passing in the random generator as an argument makes your functions pure, which helps with testing

Reply:
Thanks for your code review. I updated the random search again and outsourced the function.